src/NuGet.Protocol.Core.v3/ServiceIndexResourceV3Provider.cs

```diff
@@ -83,12 +77,12 @@ public ServiceIndexResourceV3Provider()
 
                         if (credentials == null)
                         {
-                            return null;
+                            response.EnsureSuccessStatusCode();
                         }
                     }
                     else
                     {
-                        return null;
+                        response.EnsureSuccessStatusCode();
```

Comment:
This, plus the previous EnsureSuccessStatusCode(), makes sure that this method returns only when Index.Json is successfully fetched and parsed, otherwise an exception will get thrown.